pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+package runtime
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	nethttp "net/http"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/dapr/components-contrib/contenttype"
+	contribMetadata "github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/components-contrib/pubsub"
+	diag "github.com/dapr/dapr/pkg/diagnostics"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	runtimev1pb "github.com/dapr/dapr/pkg/proto/runtime/v1"
+	"github.com/dapr/dapr/pkg/resiliency"
+	runtimePubsub "github.com/dapr/dapr/pkg/runtime/pubsub"
+)
+
+type pubsubBulkSubscribedMessage struct {
+	cloudEvents []map[string]interface{}
+	rawData     []runtimePubsub.BulkSubscribeMessageItem
+	entries     []*pubsub.BulkMessageEntry
+	data        []byte
+	topic       string
+	metadata    map[string]string
+	pubsub      string
+	path        string
+	length      int
+}
+
+type bulkSubIngressDiagnostics struct {
+	statusWiseDiag map[string]int64
+	elapsed        float64
+	retryReported  bool
+}
+
+// bulkSubscribeTopic subscribes to a topic for bulk messages and invokes subscriber app endpoint(s).
+
+// Major steps inside a bulk handler:
+//  1. Deserialize pubsub metadata and determine if rawPayload or not
+//     1.A. If any error occurs, send to DLQ if configured, else send back error for all messages
+//  2. Iterate through each message and validate entryID is NOT blank
+//     2.A. If it is a raw payload:
+//     2.A.i. Get route path, if processable
+//     2.A.ii. If route path is non-blank, generate base64 encoding of event data
+//     and set contentType, if provided, else set to "application/octet-stream"
+//     2.A.iii. Finally, form a child message to be sent to app and add it to the list of messages,
+//     to be sent to app (this list of messages is registered against correct path in an internal map)
+//     2.B. If it is NOT a raw payload (it is considered a cloud event):
+//     2.B.i. Unmarshal it into a map[string]interface{}
+//     2.B.ii. If any error while unmarshalling, send to DLQ if configured, else register error for this message
+//     2.B.iii. Check if message expired
+//     2.B.iv. Get route path, if processable
+//     2.B.v. If route path is non-blank, form a child message to be sent to app and add it to the list of messages,
+//  3. Iterate through map prepared for path vs list of messages to be sent on this path
+//     3.A. Prepare envelope for the list of messages to be sent to app on this path
+//     3.B. Send the envelope to app by invoking http endpoint
+//  4. Check if any error has occurred so far in processing for any of the message and invoke DLQ, if configured.
+//  5. Send back responses array to broker interface.
+func (a *DaprRuntime) bulkSubscribeTopic(ctx context.Context, policy resiliency.Runner,
```

Comment:
This is the main function for bulk subscription of topic in runtime. The comment above briefly defines what all is carried out here.